From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shahbaz Youssefi <syoussefi@chromium.org>
Date: Wed, 2 Sep 2026 15:16:44 -0400
Subject: Fix robust-init when OOB-copying between mips of same texture

The front-end has logic to initialize the destination texture when the
target of glCopyTexImage2D is out of bounds of the source framebuffer.
However, this is skipped during self-copies as the initialization would
destroy source(==destination)'s content before the copy.

This self-copy check is improved in this change to exclude copies
between different subresources of the same texture.

Bug: chromium:550302121
Change-Id: I565ea820d49ea8d37910247b03729468280509f0
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8347448
Reviewed-by: Geoff Lang <geofflang@chromium.org>
Commit-Queue: Shahbaz Youssefi <syoussefi@chromium.org>

diff --git a/src/libANGLE/Caps.h b/src/libANGLE/Caps.h
index 5e6c8f4704999b4037ad0103d899d09d5aa5b105..511ca0cdbdf1d599f0a4d3c70a60bc82eb43e38e 100644
--- a/src/libANGLE/Caps.h
+++ b/src/libANGLE/Caps.h
@@ -133,6 +133,10 @@ struct Limitations
     // http://anglebug.com/42266263
     bool noRasterOrderGroupWithoutAttachmentZero = false;
 
+    // Some backends don't apply robust init when glCopyTexImage2D reads out of bounds of source
+    // framebuffer, whose read attachment is a different mip of the same texture being redefined.
+    bool noRobustInitOnOOBCopyTexImageSameTexture = false;
+
     // ETC1 texture support is emulated.
     bool emulatedEtc1 = false;
 
diff --git a/src/libANGLE/Texture.cpp b/src/libANGLE/Texture.cpp
index 2c523f6c85f73078e557b8c889a37e87bd921242..7a4a0199d01ca9bafa3c4e689ebdbd09434a3943 100644
--- a/src/libANGLE/Texture.cpp
+++ b/src/libANGLE/Texture.cpp
@@ -1580,11 +1580,15 @@ angle::Result Texture::copyImage(Context *context,
         const FramebufferAttachment *sourceReadAttachment = source->getReadColorAttachment();
         Extents fbSize                                    = sourceReadAttachment->getSize();
         // Force using copySubImage when the source area is out of bounds AND
-        // we're not copying to and from the same texture
-        forceCopySubImage = ((sourceArea.x < 0) || (sourceArea.y < 0) ||
-                             ((sourceArea.x + sourceArea.width) > fbSize.width) ||
-                             ((sourceArea.y + sourceArea.height) > fbSize.height)) &&
-                            (sourceReadAttachment->getResource() != this);
+        // we're not copying to and from the same texture.  When
+        // noRobustInitOnOOBCopyTexImageSameTexture is set, copySubImage is used also when copying
+        // between different subresources of the same texture.
+        forceCopySubImage = (sourceArea.x < 0 || sourceArea.y < 0 ||
+                             sourceArea.x + sourceArea.width > fbSize.width ||
+                             sourceArea.y + sourceArea.height > fbSize.height) &&
+                            (sourceReadAttachment->getResource() != this ||
+                             (context->getLimitations().noRobustInitOnOOBCopyTexImageSameTexture &&
+                              sourceReadAttachment->getTextureImageIndex() != index));
         Rectangle clippedArea;
         if (ClipRectangle(sourceArea, Rectangle(0, 0, fbSize.width, fbSize.height), &clippedArea))
         {
diff --git a/src/libANGLE/renderer/vulkan/vk_caps_utils.cpp b/src/libANGLE/renderer/vulkan/vk_caps_utils.cpp
index 45ad108cff18312b82f5e57aa4b4cc1f9fa403be..7ec5e2d9259de1957a5996097e12dc0db8b2ab50 100644
--- a/src/libANGLE/renderer/vulkan/vk_caps_utils.cpp
+++ b/src/libANGLE/renderer/vulkan/vk_caps_utils.cpp
@@ -1517,6 +1517,10 @@ void Renderer::ensureCapsInitialized() const
     mNativeLimitations.maxBufferBytes  = static_cast<size_t>(mMaxMemoryAllocationSize);
     mNativeLimitations.maxTextureBytes = static_cast<size_t>(mMaxMemoryAllocationSize);
 
+    // Ask the front-end to handle robust-init when copying between mips of the same texture via
+    // glCopyTexImage2D.
+    mNativeLimitations.noRobustInitOnOOBCopyTexImageSameTexture = true;
+
     // Log any missing extensions required for GLES 3.2.
     LogMissingExtensionsForGLES32(mNativeExtensions);
 }
